lib/wrapeth/provider.ex

```diff
@@ -18,168 +19,192 @@ defmodule Wrapeth.Provider do
-      def eth_get_transaction_by_block_number_and_index(block, index, _args \\ []) do
+      def eth_get_transaction_by_block_number_and_index(block, index, pid \\ nil, _args \\ []) do
         call_client(:eth_get_transaction_by_block_number_and_index, [block, index])
       end
 
       @impl true
-      def eth_get_uncle_by_block_hash_and_index(hash, index, _args \\ []) do
+      def eth_get_uncle_by_block_hash_and_index(hash, index, pid \\ nil, _args \\ []) do
         call_client(:eth_get_uncle_by_block_hash_and_index, [hash, index])
       end
 
       @impl true
-      def eth_get_compilers(_args \\ []) do
-        call_client(:eth_get_compilers)
+      def eth_get_compilers(pid \\ nil, _args \\ []) do
+        call_client(:eth_get_compilers, [], pid)
       end
 
       @impl true
-      def eth_accounts(_args \\ []) do
-        call_client(:eth_accounts)
+      def eth_accounts(pid \\ nil, _args \\ []) do
+        call_client(:eth_accounts, [], pid)
       end
 
       @impl true
-      def eth_get_balance(address, block \\ "latest", _args \\ []) do
-        call_client(:eth_get_balance, [address, block])
+      def eth_get_balance(address, block \\ "latest", pid \\ nil, _args \\ []) do
+        call_client(:eth_get_balance, [address, block], pid)
       end
 
       @impl true
-      def eth_block_number(_args \\ []) do
-        call_client(:eth_block_number)
+      def eth_block_number(block \\ "latest", pid \\ nil, _args \\ []) do
+        call_client(:eth_block_number, [block], pid)
       end
 
       @impl true
-      def eth_get_block_by_number(number, full, _args \\ []) do
+      def eth_get_block_by_number(number, full, pid \\ nil, _args \\ []) do
         call_client(:eth_get_block_by_number, [number, full])
       end
 
       @impl true
-      def eth_get_transaction_count(address, block \\ "latest", _args \\ []) do
+      def eth_get_transaction_count(address, block \\ "latest", pid \\ nil, _args \\ []) do
         call_client(:eth_get_transaction_count, [address, block])
       end
 
       @impl true
-      def eth_gas_price(_args \\ []) do
-        call_client(:eth_gas_price)
+      def eth_gas_price(pid \\ nil, _args \\ []) do
+        call_client(:eth_gas_price, [], pid)
       end
 
       @impl true
-      def eth_get_transaction_receipt(hash, _args \\ []) do
+      def eth_get_transaction_receipt(hash, pid \\ nil, _args \\ []) do
         call_client(:eth_get_transaction_receipt, [hash])
       end
 
-      defp call_client(method_name, args \\ [], _arg \\ []) do
-        apply(@client_type, method_name, args ++ [[{:url, @node_url}]])
+      @impl true
+      def eth_subscribe(envent_type \\ "newHeads", pid, _args \\ []) do
+        call_client(:eth_subscribe, [envent_type], pid)
+      end
+
+      @impl true
+      def eth_unsubscribe(sub_id, pid, _args \\ []) do
+        call_client(:eth_unsubscribe, [sub_id], pid)
+      end
+
+      defp call_client(method_name, args \\ [], pid \\ nil, _args \\ []) do
```

Comment:
Do you know if we need to pass the pid as an argument? I would like to see if we could mask this and simplify everything.

Reply:
I initially attempted to find a way to avoid passing the PID as an argument because it bothers me and seems a bit messy with multiple default arguments. However, I couldn't find a solution, so I assumed that dealing with the PID is necessary for using the WebSocket. Do you have any ideas on how to handle this more efficiently?

Reply:
I think I've figured out a way to do it, so I'll give it a shot and see if it works. I noticed in a Git repo that they use a name to call the methods instead of using pid 

`GenServer.start_link(__MODULE__, :ok, name: WebSocketManager)`

Reply:
Did you consider `self()`?

Reply:
I did, but I was encountering an error related to the process attempting to call itself.

Reply:
Yes, it depends on where you were calling it.